File: frontend/src/types/telemetry.ts
// SatInsight AI — TypeScript Types: Telemetry

export interface TelemetryRecord {
  timestamp: string
  temperature: number | null
  radiation: number | null
  pressure: number | null
  battery_level: number | null
  signal_strength: number | null
  velocity: number | null
  altitude: number | null
  [key: string]: string | number | null
}

export interface DatasetMeta {
  session_id: string
  dataset_name: string
  row_count: number
  columns: string[]
  created_at: number
  has_anomalies: boolean
}

export interface DatasetLoadResponse {
  session_id: string
  dataset_name: string
  row_count: number
  columns: string[]
  warnings: string[]
  message: string
}

export interface DataPageResponse {
  session_id: string
  page: number
  page_size: number
  total_rows: number
  total_pages: number
  records: TelemetryRecord[]
}

export interface ParameterStats {
  count: number
  mean: number | null
  std: number | null
  min: number | null
  max: number | null
  q1: number | null
  median: number | null
  q3: number | null
  iqr: number | null
  missing_count: number
}

export interface DataStatsResponse {
  session_id: string
  dataset_name: string
  row_count: number
  parameters: Record<string, ParameterStats>
}

export interface DataPreviewResponse {
  session_id: string
  dataset_name: string
  row_count: number
  columns: string[]
  preview_rows: TelemetryRecord[]
  warnings: string[]
}

export const NUMERIC_PARAMS = [
  'temperature',
  'radiation',
  'pressure',
  'battery_level',
  'signal_strength',
  'velocity',
  'altitude',
] as const

export type NumericParam = (typeof NUMERIC_PARAMS)[number]

export const PARAM_UNITS: Record<NumericParam, string> = {
  temperature: '°C',
  radiation: 'mSv/h',
  pressure: 'hPa',
  battery_level: '%',
  signal_strength: 'dBm',
  velocity: 'km/s',
  altitude: 'km',
}

export const PARAM_LABELS: Record<NumericParam, string> = {
  temperature: 'Temperature',
  radiation: 'Radiation',
  pressure: 'Pressure',
  battery_level: 'Battery',
  signal_strength: 'Signal',
  velocity: 'Velocity',
  altitude: 'Altitude',
}

export const PARAM_COLORS: Record<NumericParam, string> = {
  temperature: '#f97316',
  radiation: '#a855f7',
  pressure: '#06b6d4',
  battery_level: '#22c55e',
  signal_strength: '#3b82f6',
  velocity: '#f59e0b',
  altitude: '#ec4899',
}

File: backend/api/models/telemetry.py
"""
SatInsight AI — Telemetry Pydantic Models

Response schemas used by all data-layer routes.
"""

from typing import Any, Optional
from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Session / Dataset metadata
# ---------------------------------------------------------------------------


class DatasetMeta(BaseModel):
    session_id: str
    dataset_name: str
    row_count: int
    columns: list[str]
    created_at: float
    has_anomalies: bool = False


class DatasetListResponse(BaseModel):
    sessions: list[DatasetMeta]
    total: int


# ---------------------------------------------------------------------------
# Load / Upload response
# ---------------------------------------------------------------------------


class DatasetLoadResponse(BaseModel):
    session_id: str
    dataset_name: str
    row_count: int
    columns: list[str]
    warnings: list[str] = Field(default_factory=list)
    message: str


# ---------------------------------------------------------------------------
# Paginated data records
# ---------------------------------------------------------------------------


class TelemetryRecord(BaseModel):
    """One row of satellite telemetry. All fields optional to handle partial uploads."""
    timestamp: Optional[str] = None
    temperature: Optional[float] = None
    radiation: Optional[float] = None
    pressure: Optional[float] = None
    battery_level: Optional[float] = None
    signal_strength: Optional[float] = None
    velocity: Optional[float] = None
    altitude: Optional[float] = None


class DataPageResponse(BaseModel):
    session_id: str
    page: int
    page_size: int
    total_rows: int
    total_pages: int
    records: list[dict[str, Any]]


# ---------------------------------------------------------------------------
# Statistics
# ---------------------------------------------------------------------------


class ParameterStats(BaseModel):
    count: int
    mean: Optional[float]
    std: Optional[float]
    min: Optional[float]
    max: Optional[float]
    q1: Optional[float]
    median: Optional[float]
    q3: Optional[float]
    iqr: Optional[float]
    missing_count: int


class DataStatsResponse(BaseModel):
    session_id: str
    dataset_name: str
    row_count: int
    parameters: dict[str, ParameterStats]


# ---------------------------------------------------------------------------
# Preview
# ---------------------------------------------------------------------------


class DataPreviewResponse(BaseModel):
    session_id: str
    dataset_name: str
    row_count: int
    columns: list[str]
    preview_rows: list[dict[str, Any]]   # first N rows
    warnings: list[str] = Field(default_factory=list)
